File: api/src/main/java/net/blueshell/api/validation/membership/NoExistingMembershipForUserId.java
package net.blueshell.api.validation.membership;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * Custom annotation to ensure the uniqueness of the username.
 */
@Documented
@Constraint(validatedBy = NoExistingMembershipForUserIdValidator.class)
@Target({FIELD})
@Retention(RUNTIME)
public @interface NoExistingMembershipForUserId {
    String message() default "User is already a member.";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};
}

File: api/src/main/kotlin/net/blueshell/api/domain/membership/web/validation/NoExistingMembershipForUserId.kt
package net.blueshell.api.domain.membership.web.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

/**
 * Custom annotation to ensure the uniqueness of the username.
 */
@MustBeDocumented
@Constraint(validatedBy = [NoExistingMembershipForUserIdValidator::class])
@Target(AnnotationTarget.FIELD)
@Retention(AnnotationRetention.RUNTIME)
annotation class NoExistingMembershipForUserId(
    val message: String = "User is already a member.",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = []
)
